File: scripts/generate_vehicles_complete.py
#!/usr/bin/env python3
"""
Generate complete 4-level vehicle classification from War Thunder Live API data
Combines filters structure with vehicle skin counts
"""

import json
import re
from pathlib import Path

def extract_filters_from_html(html_path):
    """Extract the filters JavaScript object from HTML response"""
    with open(html_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Find the filters object in the JavaScript
    match = re.search(r'const filters = ({.*?});', content, re.DOTALL)
    if not match:
        raise ValueError("Could not find filters object in HTML")
    
    filters_json = match.group(1)
    return json.loads(filters_json)

def load_vehicle_skins(json_path):
    """Load vehicle skins data"""
    with open(json_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def build_vehicle_hierarchy(filters, vehicle_skins):
    """Build complete 4-level vehicle hierarchy"""
    
    # Extract vehicle types
    vehicle_types = {}
    for variant in filters['vehicleType']['variants']:
        if variant['value'] != 'any':
            vehicle_types[variant['value']] = {
                'name': variant['name'],
                'count': variant.get('count', 0)
            }
    
    # Extract countries
    countries = {}
    for variant in filters['vehicleCountry']['variants']:
        if variant['value'] != 'any':
            countries[variant['value']] = {
                'name': variant['name'],
                'count': variant.get('count', 0)
            }
    
    # Extract vehicle classes with dependencies
    vehicle_classes = {}
    for variant in filters['vehicleClass']['variants']:
        if variant.get('value') and not variant.get('separator'):
            dep = variant.get('dep', {})
            vehicle_classes[variant['value']] = {
                'name': variant['name'],
                'count': variant.get('count', 0),
                'vehicleType': dep.get('vehicleType', [])
            }
    
    # Extract vehicles with full dependencies
    vehicles = {}
    for variant in filters['vehicle']['variants']:
        if variant.get('value') and variant['value'] != 'any' and not variant.get('separator'):
            dep = variant.get('dep', {})
            vehicle_id = variant['value']
            
            # Get count from vehicle_skins if available
            count = vehicle_skins.get(vehicle_id, {}).get('count', variant.get('count', 0))
            
            vehicles[vehicle_id] = {
                'name': variant['name'],
                'count': count,
                'vehicleCountry': dep.get('vehicleCountry', []),
                'vehicleType': dep.get('vehicleType', []),
                'vehicleClass': dep.get('vehicleClass', [])
            }
    
    return {
        'vehicleTypes': vehicle_types,
        'countries': countries,
        'vehicleClasses': vehicle_classes,
        'vehicles': vehicles
    }

def main():
    # Paths
    html_path = Path('api_response_head.html')
    skins_path = Path('wt_vehicle_skins.json')
    output_path = Path('server/data/vehicles_complete.json')
    
    print("Loading data...")
    filters = extract_filters_from_html(html_path)
    vehicle_skins = load_vehicle_skins(skins_path)
    
    print("Building vehicle hierarchy...")
    hierarchy = build_vehicle_hierarchy(filters, vehicle_skins)
    
    print(f"Found:")
    print(f"  - {len(hierarchy['vehicleTypes'])} vehicle types")
    print(f"  - {len(hierarchy['countries'])} countries")
    print(f"  - {len(hierarchy['vehicleClasses'])} vehicle classes")
    print(f"  - {len(hierarchy['vehicles'])} vehicles")
    
    # Save to file
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(hierarchy, f, ensure_ascii=False, indent=2)
    
    print(f"\nSaved to {output_path}")
    
    # Print sample
    print("\nSample vehicle (first one):")
    first_vehicle_id = list(hierarchy['vehicles'].keys())[0]
    print(f"  ID: {first_vehicle_id}")
    print(f"  Data: {json.dumps(hierarchy['vehicles'][first_vehicle_id], ensure_ascii=False, indent=4)}")

if __name__ == '__main__':
    main()

File: scripts/generate_vehicles_complete.js
#!/usr/bin/env node
/**
 * Generate complete 4-level vehicle classification from War Thunder Live API data
 * Combines filters structure with vehicle skin counts
 */

const fs = require('fs');
const path = require('path');

function extractFiltersFromHtml(htmlPath) {
    const content = fs.readFileSync(htmlPath, 'utf-8');
    
    // Find the filters object in the JavaScript
    const match = content.match(/const filters = ({.*?});/s);
    if (!match) {
        throw new Error('Could not find filters object in HTML');
    }
    
    return JSON.parse(match[1]);
}

function loadVehicleSkins(jsonPath) {
    return JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
}

function buildVehicleHierarchy(filters, vehicleSkins) {
    // Extract vehicle types
    const vehicleTypes = {};
    for (const variant of filters.vehicleType.variants) {
        if (variant.value !== 'any') {
            vehicleTypes[variant.value] = {
                name: variant.name,
                count: variant.count || 0
            };
        }
    }
    
    // Extract countries
    const countries = {};
    for (const variant of filters.vehicleCountry.variants) {
        if (variant.value !== 'any') {
            countries[variant.value] = {
                name: variant.name,
                count: variant.count || 0
            };
        }
    }
    
    // Extract vehicle classes with dependencies
    const vehicleClasses = {};
    for (const variant of filters.vehicleClass.variants) {
        if (variant.value && !variant.separator) {
            const dep = variant.dep || {};
            vehicleClasses[variant.value] = {
                name: variant.name,
                count: variant.count || 0,
                vehicleType: dep.vehicleType || []
            };
        }
    }
    
    // Extract vehicles with full dependencies
    const vehicles = {};
    for (const variant of filters.vehicle.variants) {
        if (variant.value && variant.value !== 'any' && !variant.separator) {
            const dep = variant.dep || {};
            const vehicleId = variant.value;
            
            // Get count from vehicleSkins if available
            const count = vehicleSkins[vehicleId]?.count || variant.count || 0;
            
            vehicles[vehicleId] = {
                name: variant.name,
                count: count,
                vehicleCountry: dep.vehicleCountry || [],
                vehicleType: dep.vehicleType || [],
                vehicleClass: dep.vehicleClass || []
            };
        }
    }
    
    return {
        vehicleTypes,
        countries,
        vehicleClasses,
        vehicles
    };
}

function main() {
    const htmlPath = path.join(__dirname, '..', 'api_response_head.html');
    const skinsPath = path.join(__dirname, '..', 'wt_vehicle_skins.json');
    const outputPath = path.join(__dirname, '..', 'server', 'data', 'vehicles_complete.json');
    
    console.log('Loading data...');
    const filters = extractFiltersFromHtml(htmlPath);
    const vehicleSkins = loadVehicleSkins(skinsPath);
    
    console.log('Building vehicle hierarchy...');
    const hierarchy = buildVehicleHierarchy(filters, vehicleSkins);
    
    console.log(`Found:`);
    console.log(`  - ${Object.keys(hierarchy.vehicleTypes).length} vehicle types`);
    console.log(`  - ${Object.keys(hierarchy.countries).length} countries`);
    console.log(`  - ${Object.keys(hierarchy.vehicleClasses).length} vehicle classes`);
    console.log(`  - ${Object.keys(hierarchy.vehicles).length} vehicles`);
    
    // Save to file
    const outputDir = path.dirname(outputPath);
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    fs.writeFileSync(outputPath, JSON.stringify(hierarchy, null, 2), 'utf-8');
    
    console.log(`\nSaved to ${outputPath}`);
    
    // Print sample
    const firstVehicleId = Object.keys(hierarchy.vehicles)[0];
    console.log('\nSample vehicle (first one):');
    console.log(`  ID: ${firstVehicleId}`);
    console.log(`  Data: ${JSON.stringify(hierarchy.vehicles[firstVehicleId], null, 4)}`);
}

main();
